doc/contributing.rst

```diff
@@ -1,221 +1,760 @@
+
+Other commands that you will undoubtedly need relate to `branches`_. Branches
+represent multiple copies of the codebase *within a local clone or remote
+repo*. Branches are typically used to experiment with new features while still
+keeping a clean, working copy of the original codebase that you can switch back
+to at any time. The default branch of any repo is always called ``master``, and
+it is recommended that you reserve the ``master`` branch to be that clean copy
+of the working "upstream" codebase. Therefore, if you want to add a new
+feature, you should first synchronize your local ``master`` branch with the
+``upstream`` repository, then create a new branch based off of ``master`` and
+`check it out`_ so that any changes you make will exist on that new branch
+(instead of on ``master``):
+
+.. code-block:: console
+
+    $ git checkout master            # switch to master branch
+    $ git fetch upstream             # get the current state of the upstream repo
+    $ git merge upstream/master      # synchronize local master branch with upstream master branch
+    $ git checkout -b new-feature-x  # create branch "new-feature-x" and check it out
+
+.. note::
+
+    You can save some typing by using ``git pull upstream/master`` to replace
+    the ``fetch`` and ``merge`` lines above.
+
+Now that you're on a new branch, you can fix a bug or add a new feature, add a
+test, update the documentation, etc. When you're done, it's time to organize
+your changes into a series of `commits`_. Commits are like snapshots of the
+repository — actually, more like a description of what has to change to get
+from the most recent snapshot to the current snapshot.
+
+Git knows that people often work on multiple changes in multiple files all at
+once, but that ultimately they should separate those changes into sets of
+related changes that are grouped together based on common goals (so that it's
+easier for their colleagues to understand and review the changes). For example,
+you might want to group all the code changes together in one commit, put new
+unit tests in another commit, and changes to the documentation in a third
+commit.  Git makes this easy(ish) with something called the `stage`_ (or
+*staging area*). After you've made some changes to the codebase, you'll have
+what git calls "unstaged changes", which will show up with the `status`_
+command:
+
+.. code-block:: console
+
+    $ # see what state the local copy of the codebase is in
+    $ git status
+
+Those unstaged changes can be `added`_ to the stage one by one, by either
+adding a whole file's worth of changes, or by adding only certain lines
+interactively:
+
+.. code-block:: console
+
+    $ # add a whole file's worth of changes
+    $ # (same command works to add a completely new file):
+    $ git add mne/some_file.py
+    $ # enter interactive staging mode, to add only portions of the file:
+    $ git add -p mne/viz/some_other_file.py
```

Comment:
This interactive bit seems like overkill / expert territory. 

FWIW 99% of the time I just do `git commit -am "whatever"`. Might be simpler than telling people to do it one by one.

Reply:
When I was starting with git learning about `-p` actually made my life easier - beginnings are often messy and its good to know how to clean up. So it would be nice to still keep this information somwhere.

Reply:
I've left this in because I agree with @mmagnuski, `add -p` is really useful for beginners I think.